logpyle/__init__.py

```diff
@@ -638,18 +634,72 @@ def __init__(self, filename: Optional[str] = None, mode: str = "r",
 
         # {{{ warnings/logging capture
 
-        self.warning_data: List[_LogWarningInfo] = []
+        warning_data: List[_LogWarningInfo] = []
+        self.warning_data = warning_data
         self.old_showwarning: Optional[Callable[..., Any]] = None
         if capture_warnings and self.mode[0] == "w":
             self.capture_warnings(True)
 
-        self.logging_data: List[_LogWarningInfo] = []
+        logging_data: List[_LogWarningInfo] = []
+        self.logging_data = logging_data
         self.logging_handler: Optional[logging.Handler] = None
         if capture_logging and self.mode[0] == "w":
             self.capture_logging(True)
 
         # }}}
 
+        # {{{ atexit handling
+
+        def atexit_close() -> None:
+            # Note: This function can not reference 'self', otherwise
+            # atexit will hold a reference to the LogManager instance,
+            # keeping that instance alive.
+            rank = mpi_comm.rank if mpi_comm else 0
+
+            def atexit_save_logging() -> None:
+                nonlocal logging_data
+                for log in logging_data:
+                    db_conn.execute(
+                        "insert into logging values (?,?,?,?,?,?,?)",
+                        (rank, log.tick_count, log.time,
+                        log.category, log.message, log.filename,
+                        log.lineno))
+
+                logging_data = []
+
+            def atexit_save_warnings() -> None:
+                nonlocal warning_data
+                for w in warning_data:
+                    db_conn.execute(
+                        "insert into warnings values (?,?,?,?,?,?,?)",
+                        (rank, w.tick_count, w.time, w.message,
+                            w.category, w.filename, w.lineno))
+
+                warning_data = []
```

Comment:
This is duplicated code---factor it out?

Reply:
I remove d the code in c4ec70a.